src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs

```diff
@@ -48,6 +48,80 @@ static class BitOperations
             19, 27, 23, 06, 26, 05, 04, 31
         };
 
+        /// <summary>
+        /// Evaluate whether a given integral value is a power of 2.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsPow2(int value)
+        {
+            if (Popcnt.IsSupported)
+            {
+                // 1 set bit means number is a power of 2 unless it is the sign bit, so get rid of that
+                // Cast to unsigned type so logic shift not signed arithmetic shift
+                return PopCount((uint)value << 1) == 1;
```

Comment:
`(int)0x80000001` is not a power of 2, so you can't just discard the sign bit.

Reply:
Good point, it doesn't match software fallback behavior (add more test cases)

Reply:
It actually matches the bug in the `SoftwareFallback((uint)value << 1)` function call. 😉

Reply:
ah, indeed, I didn't notice `<< 1`

Reply:
Oops sorry - will add a test for that and fix the fallback. I think the best way to do it in the accelerated route is
```c#
uint unsigned = (uint)value;
uint withoutSign = unsigned << 1;
uint mask = unsigned >> 31;
return PopCount(withoutSign) & mask == 1;
```

which is probably cheaper than a branch

Reply:
> ```cs
> return PopCount(withoutSign) & ~mask == 1;
> ```
That will not compile.

Reply:
It was just a rough impl i wrote on my phone that gets the idea across 🙂

Reply:
If you're checking the sign bit separately, then you don't need the `<< 1` before `PopCount`.

Reply:
I was trying to avoid the branch introduced by explicitly checking the sign bit

Reply:
I meant, if you are using `>> 31` and `& ~mask` to ensure that `IsPow2(value)` returns `false` when `value < 0`, then it does not matter what `PopCount` returns for those values, so the `<< 1` is not necessary.

Is the code with `PopCount` actually faster than the software fallback, in which the `value >= 0` check can be combined with the `value != 0` check that is necessary anyway?

Reply:
I'm on Skylake, where popcount isn't as fast, and it seems branchless popcount is still generally faster than branch + popcount or the software fallback. Different should be even bigger on Ryzen where popcount is faster

Reply:
Could you provide some numbers and the benchmark you are running?

Reply:
Why isn't this just `(value > 0) && IsPow2((uint)value)`?

Negative values generally aren't considered to be powers of 2, since you can't raise either `2` or `-2` to `n` to get them.

Reply:
`IsPow2((uint)value)` almost works, except that `-2147483648` would return true.